### ISSUE TITLE
org-block-begin-line

### ISSUE THREAD

**Issue by bestlem** [protesilaos/modus-themes#32]
I notice that in the current modus-themes you are missing an inherit.
```
  `(org-block-begin-line ((,class :inherit modus-themes-fixed-pitch
                                    ,@(modus-themes--org-block-delim
                                       bg-dim fg-special-cold
                                       bg-alt fg-alt))))
```

But in org-mode that comes with emacs 28.1 org-block-begin-line inherits from org-meta-line
```
(defface org-block-begin-line '((t (:inherit org-meta-line)))
  "Face used for the line delimiting the begin of source blocks."
  :group 'org-faces)
```

**Comment by protesilaos** [comment]
Hi @bestlem!

> But in org-mode that comes with emacs 28.1 org-block-begin-line
> inherits from org-meta-line

You mean there has been some kind of incompatibility?  If so I will
address it right away.

The `:inherit` we have there is to enforce the user option
`modus-themes-mixed-fonts`.  When that variable is `t`, all
spacing-sensitive elements use the `fixed-pitch` face to stay monospaced
even if the `default` face is proportionately spaced (e.g. when you use
`M-x variable-pitch-mode`).

**Comment by bestlem** [comment]
In org-mode code 

The face org-block-begin-line inherits org-meta
The face org-meta inherits (fixed-pitch font-lock-comment-face))

So I think org-meta is the one that you would need to modify.

On a connected point I call variable-pitch-mode to allow org-mode text to be variable font and org-block to be fixed font. This mode is part of emacs so I am not clear what you are adding by modus-themes-mixed-fonts.

Also your example in https://protesilaos.com/emacs/modus-themes#h:defcf4fc-8fa8-4c29-b12e-7119582cc929 I think only works as the height for variable-pitch is 1.0

What variable-pitch-mode does is change the default font to the variable-pitch. Thus all heights are now relative to the variable pitch's height. I has variable-pitch height of 1.5 and so fixed font in org-mode was bigger than in plain code mode.
I think both fixed and variable font pitches need to be absoult i.e. whole numbers

**Comment by protesilaos** [comment]
> The face org-block-begin-line inherits org-meta The face org-meta inherits (fixed-pitch font-lock-comment-face))
>
> So I think org-meta is the one that you would need to modify.

I don't have that face in my build of Emacs.  a search for "org-meta" returns the following:

```
org-meta-line           a       Face for meta lines starting with "#+".
org-metaup              c       Move subtree up or move table row up.
org-metadown            c       Move subtree down or move table row down.
org-metaleft            c       Promote heading, list item at point or move table column left.
org-metaright           c       Demote heading, list item at point or move table column right.
org-meta-return         c       Insert a new heading or wrap a region in a table.
org-metaup-hook         v       Hook for functions attaching themselves to ‘M-up’.
org-metadown-hook       v       Hook for functions attaching themselves to ‘M-down’.
org-metaleft-hook       v       Hook for functions attaching themselves to ‘M-left’.
org-metaright-hook      v       Hook for functions attaching themselves to ‘M-right’.
org-metareturn-hook     v       Hook for functions attaching themselves to ‘M-RET’.
```

Maybe you mean `org-meta-line`?  If so, we do configure it in the same spirit as you describe:

```elisp
`(org-meta-line ((,class :inherit (shadow modus-themes-fixed-pitch))))
```

We do not inherit from `font-lock-comment-face` because those are not "comments" per se and we need the flexibility to differentiate between the two.

> On a connected point I call variable-pitch-mode to allow org-mode text to be variable font and org-block to be fixed font. This mode is part of emacs so I am not clear what you are adding by modus-themes-mixed-fonts.

If `modus-themes-mixed-fonts` is nil, `M-x variable-pitch-mode` will make everything proportionately spaced, including tables and code blocks.  In some older versions of Org, certain faces were inheriting from `fixed-pitch` (I made that change), but we discovered it was not a good default: it is removed upstream.

To enforce `fixed-pitch` where it needs to be (e.g. tables and source blocks), you need to set `modus-themes-mixed-fonts` to `t`.

> I think both fixed and variable font pitches need to be absoult i.e. whole number

Absolute values break `text-scale-adjust`, depending on the specifics.  Though it ultimately is up to the user to set font families and font heights: the themes do not make any changes on that front.

For a package that handles faces+fonts, consider `fontaine` (I am the author).  Here is its mirror on GitHub: <https://github.com/protesilaos/fontaine>.

**Comment by bestlem** [comment]
The code is in org-mode in from gnu ELPA 

```
(defface org-block-begin-line '((t (:inherit org-meta-line)))
  "Face used for the line delimiting the begin of source blocks."
  :group 'org-faces)
```
So modus-themes is overwriting an inheritence.

**Comment by protesilaos** [comment]
> However the issue I have is that variable-pitch-mode changes the default and all relative sizes are now relative to the variable font. This makes the font of code in org-src blocks different to that in plain code buffers unless variable-font is the same height as default.

This is how `variable-pitch-mode` works.  It is not up to the theme to control that behaviour.  As I noted before, the user is free to set the `variable-pitch` face to an absolute `:height` value: the themes do not touch that face and generally do not hardcode typographic preferences.

> ie default is say :height 140 and I need variable to be 1.2 so fixed font height is 1.0 in code modes and 1/1.2 in org-mode

In that case you are better off with absolute height values instead of 1.2.  This, however, will break `text-scale-adjust`.  Again, that is not the theme's fault.

> I want to change fonts per buffer/window or even by mode.

I think it is only possible per frame.  Unless you use face remapping, which is buffer-local.  Though that is not a trivial setup and needs to be handled with care.

> So modus-themes is overwriting an inheritence.

Indeed.  There is a good reason for that, because the begin/end lines of code blocks have different requirements than the generic `org-meta-line`.  For example, they need to have a background colour that stretches to the edge of the window when the user opts in to `(setq modus-themes-org-blocks 'gray-background)`.

Do you have a problem with this arrangement?

**Comment by protesilaos** [comment]
I see.  In this case I am closing the issue.  Thanks!
